jerry-core/ecma/builtin-objects/ecma-builtins.c

```diff
@@ -949,6 +949,37 @@ ecma_builtin_dispatch_routine (ecma_builtin_id_t builtin_object_id, /**< built-i
                                ecma_length_t arguments_list_len) /**< length of arguments' list */
 {
   JERRY_ASSERT (builtin_object_id < ECMA_BUILTIN_ID__COUNT);
+
+  ecma_value_t padded_arguments_list_p[3] = { ECMA_VALUE_UNDEFINED, ECMA_VALUE_UNDEFINED, ECMA_VALUE_UNDEFINED };
+
+  if (arguments_list_len <= 3)
+  {
+    switch (arguments_list_len)
+    {
+      case 3:
+      {
+        padded_arguments_list_p[2] = arguments_list_p[2];
+        /* FALLTHRU */
+      }
```

Comment:
I think that this case is unnecessary and the outer `if` should only check for `arguments_list_len <= 2`. If `arguments_list_len == 3`, then there is no need for `padded_arguments_list_p` as `arguments_list_p` already has 3 elements.

Reply:
I cannot agree with that. We have routines with 3 arguments like `Object.defineProperty`.

```c
/**
 * The Object object's 'defineProperty' routine
 */
static ecma_value_t
ecma_builtin_object_object_define_property (ecma_value_t this_arg, /**< 'this' argument */
                                            ecma_value_t arg1, /**< routine's first argument */
                                            ecma_value_t arg2, /**< routine's second argument */
                                            ecma_value_t arg3) /**< routine's third argument */
```

These 3 arguments must came from the `padded_arguments_list_p` which arguments must be `undefined` or `arguments_list_p[i]` where `i>= 0 , i<3`

Reply:
Let's break down to cases:
- `arguments_list_len == 0`: we need `padded_arguments_list_p` and all its elements should be `ECMA_VALUE_UNDEFINED`. `arguments_list_p` will be replaced by `padded_arguments_list_p` and `ecma_builtin_routines[builtin_object_id]` will be called.
- `arguments_list_len == 1`: we need `padded_arguments_list_p`, its 0th element should be copied from `arguments_list_p` and the following two elements should be `ECMA_VALUE_UNDEFINED`. `arguments_list_p` will be replaced by `padded_arguments_list_p` and `ecma_builtin_routines[builtin_object_id]` will be called.
- `arguments_list_len == 2`: we need `padded_arguments_list_p`, its 0th and 1st elements should be copied from `arguments_list_p` and the last element should be `ECMA_VALUE_UNDEFINED`. `arguments_list_p` will be replaced by `padded_arguments_list_p` and `ecma_builtin_routines[builtin_object_id]` will be called.
- `arguments_list_len == 3`: we _do not_ need `padded_arguments_list_p` as `arguments_list_p` already has 3 elements so it can be used as is to call `ecma_builtin_routines[builtin_object_id]`.

Reply:
Indeed, the most trivial case tricked me a bit! :) Thanks for the clarification.